dao/tenant_dao.go

```diff
@@ -68,3 +77,141 @@ func (t *tenantDaoImpl) TenantByIdentity(id *identity.Identity) (*m.Tenant, erro
 
 	return &tenant, nil
 }
+
+func (t *tenantDaoImpl) GetUntranslatedTenants() ([]m.Tenant, error) {
+	var tenants []m.Tenant
+
+	err := DB.Debug().
+		Model(&m.Tenant{}).
+		Where(`("external_tenant" IS NOT NULL OR "external_tenant" = '') AND ("org_id" IS NULL OR "org_id" = '')`).
+		Find(&tenants).
+		Error
+
+	if err != nil {
+		return nil, err
+	}
+
+	return tenants, err
+}
+
+func (t *tenantDaoImpl) TranslateTenants() (int64, uint64, uint64, []m.TenantTranslation, error) {
+	// The where condition to fetch only the untranslated tenants.
+	untranslatedTenantsWhereCondition := `("external_tenant" IS NOT NULL OR "external_tenant" = '') AND ("org_id" IS NULL OR "org_id" = '')`
+
+	// Count the total number of translatable tenants.
+	var translatableTenants int64
+	// Count the translation operations.
+	var translatedTenants, untranslatedTenants uint64
+	// Keep track of each of the tenant translations.
+	var tenantTranslations []m.TenantTranslation
+
+	// Run everything inside a transaction to avoid inconsistencies. This way if new tenants are added/removed while
+	// the translation is in progress, all the counts and the results will still match.
+	err := DB.Debug().Transaction(func(tx *gorm.DB) error {
+		// Get the count to be able to process the tenants in batches of 100 tenants.
+		err := DB.Debug().
+			Model(&m.Tenant{}).
+			Where(untranslatedTenantsWhereCondition).
+			Count(&translatableTenants).
+			Error
+
+		if err != nil {
+			return fmt.Errorf("unable to count untranslated tenants: %w", err)
+		}
+
+		translator := tenantid.NewTranslator(config.Get().TenantTranslatorUrl)
+		for i := int64(0); i < translatableTenants; i += 100 {
+			// Grab all the EBS account numbers to be processed.
+			var ebsAccountNumbers []string
+			err := DB.
+				Debug().
+				Model(&m.Tenant{}).
+				Where(untranslatedTenantsWhereCondition).
+				Pluck("external_tenant", &ebsAccountNumbers).
+				Offset(int(i)).
+				Error
+
+			if err != nil {
+				return fmt.Errorf("unable to fetch the tenants to translate them: %w", err)
+			}
+
+			// Batch process 100 EBS account numbers.
+			results, err := translator.EANsToOrgIDs(context.Background(), ebsAccountNumbers)
+			if err != nil {
+				return fmt.Errorf("error when calling the translation service: %w", err)
+			}
+
+			for _, result := range results {
+				var tenantTranslation = m.TenantTranslation{}
+
+				// Did the translation service return an error?
+				if result.Err != nil {
+					logging.Log.WithField("external_tenant", *result.EAN).Errorf(`could not translate to "org_id": %s`, err)
+
+					tenantTranslation.ExternalTenant = *result.EAN
+					tenantTranslation.Error = err
+					tenantTranslations = append(tenantTranslations, tenantTranslation)
+
+					untranslatedTenants++
+					continue
+				}
+
+				// Grab the updated tenant for the "translated tenant" struct.
+				var tenant m.Tenant
+				// Update the tenant with the translated "orgId".
+				dbResult := DB.Debug().
```

Comment:
I did not play with transaction in gorm but should not be here `tx` instead of `DB.Debug()` ?

Reply:
Ooof you're right, yes. I wrapped everything on a transaction and then I forgo to use it :joy: Thanks!